### ISSUE TITLE
PIPELINE SEQUENCE TESTING: query_semiology pipeline needs to be compared to query_lateralisation pipeline

### ISSUE THREAD

**Issue by thenineteen** [thenineteen/Semiology-Visualisation-Tool#94]
currently in semiology.py I had suggested to @fepegar to run this pipeline when no lateralising data, (which I've only just updated the if statements)

```
                if all_combined_gifs is None:
                    # then either no lateralising pt data, or empty lat column
                    # run manual pipeline:
                    pivot_result = melt_then_pivot_query(
                        mega_analysis_df,
                        query_semiology_result,
                        self.term,
                    )
                    all_combined_gifs = pivot_result_to_one_map(
                        pivot_result,
                        suppress_prints=True,
                        map_df_dict=map_df_dict,
                    )
        else:
            # else: query semiology is none, or both localising and lateralising are zero:
            print('No such semiology found.')
            return
        return all_combined_gifs

```

specifically, the all_combined_gifs returned by the manual pipeline is likely to be different to that from QUERY_LATERALISATION and I need to check this and document

**Comment by thenineteen** [comment]
`semiology.py` line 187


```
if all_combined_gifs is None:
                    # Either no lateralising pt data, or empty lat column
                    # Run manual pipeline:
                    pivot_result = melt_then_pivot_query(
                        mega_analysis_df,
                        query_semiology_result,
                        self.term,
                    )
                    all_combined_gifs = pivot_result_to_one_map(
                        pivot_result,
                        map_df_dict=map_df_dict,
                    )
```

**Comment by thenineteen** [comment]
However, by the time I wrote `QUERY_LATERALISATION`, I thought we should get the gif parcellations first (mapping) then melt and pivot and do any QuantileTransforming afterwards

so Q_L runs `pivot_result_to_one_map()` but does not call `melt_then_pivot_query()`

This difference in pipelines is what I need to explore

**Comment by thenineteen** [comment]
probably best way to do this is dummy data to check output of a semiology "lateralises semio A" with "no lateralising datapoints semio B" while both map to the same exact localisations with the same exact values.

this tests if the outputs are equal.

This would work as in Q_L there is this:
```
    try:
        Lat = inspect_result['Lateralising']
        logging.debug(f'Lateralisation based on: {Lat.sum()} datapoints')
    except KeyError:
        print('No Lateralising values found for this query of the database.')
        return
```

**Comment by thenineteen** [comment]
> probably best way to do this is dummy data to check output of a semiology "lateralises semio A" with "no lateralising datapoints semio B" while both map to the same exact localisations with the same exact values.
> 
> this tests if the outputs are equal.
> 
> This would work as in Q_L there is this:
> 
> ```
>     try:
>         Lat = inspect_result['Lateralising']
>         logging.debug(f'Lateralisation based on: {Lat.sum()} datapoints')
>     except KeyError:
>         print('No Lateralising values found for this query of the database.')
>         return
> ```

this is out of date. But the idea of the test explained here is implemented in 
`test_compare_pipelines1`

and

`test_compare_pipelines3`

see docstrings.

No changes were necessary - tests all pass.

**Comment by thenineteen** [comment]
9cac76e introduced this in `Q_L` which also captures Lateralising==0. This seems to have been missed out resulting in the manual pipeline (without lateralisation) not happening:

```python
try:
        num_QL_lat = inspect_result['Lateralising'].sum()
        if num_QL_lat > 0:
            logging.debug(
                f'\n\nLateralisation based on: {num_QL_lat.sum()} datapoints')
        else:
            num_QL_lat = None
            return None, None, None, None, None, None, None
  

```

After it has been introduced, we now see the pipelines **are** _slightly_ different.


`test_compare_pipelines2` shows GIF parcellations 105 and 106 get a 6.0 in 50:50 lateralising QL, whereas the manual pipeline QS (`melt_then_pivot_query` then `pivot_result_to_one_map`) gives them a 4.0. The rest of 'pt #s' are the same (i.e. only 2 out of 52 GIF parcellations are different in this test)

**Comment by thenineteen** [comment]
GIFs `105` and `106` from `test_compare_pipelines2` are ant orbital gyrus. Should be 2 from `Ant OF`. 

If no Hierarchy, would expect it to be 4 or 8 (2 from `Ant OF`, 2 from `Orbito-frontal BA10` and 4 from `FL`)

**Comment by thenineteen** [comment]
- [ ] review this in `Q_L`

```python
if Right == Left:
            # no point as this is 50:50 as it already is, so skip
            # before continuing, ensure there is a all_combined_gifs...
            # e.g. for blink there isn't as the first row is 50:50...
            # ... and all future codes fail
            if all_combined_gifs is None:
                all_combined_gifs = row_to_one_map
            elif all_combined_gifs is not None:
                all_combined_gifs = pd.concat(
                    [all_combined_gifs, row_to_one_map], join='outer', sort=False)
            continue
```

**Comment by thenineteen** [comment]
> GIFs `105` and `106` from `test_compare_pipelines2` are ant orbital gyrus. Should be 2 from `Ant OF`.
> 
> If no Hierarchy, would expect it to be 4 or 8 (2 from `Ant OF`, 2 from `Orbito-frontal BA10` and 4 from `FL`)
> 
> ![image](https://user-images.githubusercontent.com/10786099/92272136-e75df180-eee0-11ea-99e9-11dae590fff3.png)
> 
> ![image](https://user-images.githubusercontent.com/10786099/92272208-08264700-eee1-11ea-87c8-48ba3519ceff.png)

well this clearly shows the problem should be in the `hierarchy_dictionary` module as the figures don't match (2, 2, 2 for ant OF, orbito-frontal... and FL whereas the outputs are 4.0 or 6.0).

hierarchy_dictionary is missing the `Orbito-frontal (BA 10, 11, 12/47) (basal = gyrus rectus and OFC)` as a value for key `FL`.

**Comment by thenineteen** [comment]
debug reveals `hierarchy_dictionaries` had a different spelling for `Medial Frontal (include medial premotor and its constituents as its subsets)` and so 2 were not being deducted from FL during the `graunlar=True` option.

**Comment by thenineteen** [comment]
both tests still fail

`test_compare_pipelines2`
but now the manual pipeline in this test (QS) doesn't have any 105 or 106 GIFs and so the shapes are also different.
Whereas the QL pipeline for 50:50 lateralising data shows 105 and 106 GIFs `pt #s` both a 2.0 as expected.

**Comment by thenineteen** [comment]
Aha! there is a superfluous function called `anatomical_regions()` in the module `group_columns`. This filters the localisations based on what was present in the database - it turns out Ant OF wasn't, and this is therefore excluded when present in the testing.

**Comment by thenineteen** [comment]
the next pipeline3 seems to fail due to float imprecise comparisons introduced when removing forcing 'pt #s' to int32

**Comment by thenineteen** [comment]
yikes, error was propagated during dummy_data entry!

**Comment by thenineteen** [comment]
fixed in various commits, mainly 

- `dummy_data` correction f547a23 then c27cb28

- `hierarchy_dictionary` corrections 20aebe6 for medial frontal

- not forcing int32 during testing bc84bb8

- `anatomical_regions()` in `group_columns` fixed 9468350
